### ISSUE TITLE
loading saved model for testing

### ISSUE THREAD

**Issue by drforester** [facebookresearch/Adversarial-Continual-Learning#7]
It's wonderful that this code was released to accompany the paper; thanks!
I have trained on the 20-task CIFAR-100 dataset and the models are saved to src/checkpoints. I am now attempting to call the test method in acl.py without first calling the train method like was done during training. I manually set the checkpoint to correspond to the dir of the saved model. I changed the relevant parts of the run() function in main.py to look like this image:
![run](https://user-images.githubusercontent.com/60823216/83298484-2a117e80-a1ba-11ea-9124-d4b3e3647923.png)
The first task returns a reasonable result I suppose (I only trained the network for 2 epochs each task). But when test is called on the data from the next task, I get the errors in the next image:
![errors_0](https://user-images.githubusercontent.com/60823216/83298680-7bba0900-a1ba-11ea-81a4-09b970883704.png)
There is more output, so here's the final part:
![errors_1](https://user-images.githubusercontent.com/60823216/83298752-98564100-a1ba-11ea-98a4-eedc805b9a01.png)
Is there a way to do what I am attempting without substantially changing the code? I know that this is not a shortcoming with the code, I just couldn't help trying some tests with a view to eventually training on my own dataset and using the model for inference.

**Comment by SaynaEbrahimi** [comment]
Thank you for your interest in our work. 

First, let me explain why you got those errors. I think there is a discrepancy between the discriminator in the final model and test model because the size of the discriminator's final layer increases during training at arrival of each task. Hence you can only use `test()` method if you remove the accuracy calculation for the discriminator.  

In order to test a model trained up to task T on a specific task (for example task K) you need to load both of the models you trained up to task K and T and then replace the shared module of the model trained up to task K with that from the model trained up to task T. 

I made the following code snippets for you which does what I said above so you can use them directly following these steps:

1. Put `test_trained_model()` in the ``main.py`` file and call it in `if __name__ == '__main__':` instead of `main(args)`.

```python
def test_trained_model(args, final_model_id):
    args.seed = 0

    print('Instantiate data generators and model...')
    dataloader = datagenerator.DatasetGen(args)
    args.taskcla, args.inputsize = dataloader.taskcla, dataloader.inputsize
    if args.experiment == 'multidatasets': args.lrs = dataloader.lrs

    def get_model(final_model_id, test_data_id):
        # Load the test model
        test_net = network.Net(args)
        checkpoint_test = torch.load(os.path.join(args.checkpoint, 'model_{}.pth.tar'.format(test_data_id)))
        test_net.load_state_dict(checkpoint_test['model_state_dict'])

        # Load your final trained model
        net = network.Net(args)
        checkpoint = torch.load(os.path.join(args.checkpoint, 'model_{}.pth.tar'.format(final_model_id)))
        net.load_state_dict(checkpoint['model_state_dict'])

        # # Change the shared module with the final model's shared module
        final_shared = deepcopy(net.shared.state_dict())
        test_net.shared.load_state_dict(final_shared)
        test_net = test_net.to(args.device)

        return test_net

    for t,ncla in args.taskcla:
        print('*'*250)
        dataset = dataloader.get(t)
        print(' '*105, 'Dataset {:2d} ({:s})'.format(t+1,dataset[t]['name']))
        print('*'*250)

        # Model
        test_model = get_model(final_model_id, test_data_id=t)

        # Approach
        appr = approach(test_model, args, network=network)

        # Test
        test_res = appr.inference(dataset[t]['test'], t, model=test_model)

        print('>>> Test on task {:2d} - {:15s}: loss={:.3f}, acc={:5.4f}% <<<'.format(t, dataset[t]['name'],
                                                                                          test_res['loss_t'],
                                                                                          test_res['acc_t']))
```

2. The above function calls for `inference()` in ACL class which you can find below. Insert this into your ACL class in `acl.py`.

```python
    def inference(self, data_loader, task_id, model):
        loss_a, loss_t, loss_d, loss_total = 0, 0, 0, 0
        correct_d, correct_t = 0, 0
        num = 0
        batch = 0
        model.eval()
        res = {}
        with torch.no_grad():
            for batch, (data, target, tt, td) in enumerate(data_loader):
                x = data.to(device=self.device)
                y = target.to(device=self.device, dtype=torch.long)
                tt = tt.to(device=self.device)

                # Forward
                output = model.forward(x, x, tt, task_id)
                shared_out, task_out = model.get_encoded_ftrs(x, x, task_id)

                _, pred = output.max(1)
                correct_t += pred.eq(y.view_as(pred)).sum().item()

                if self.diff == 'yes':
                    diff_loss = self.diff_loss(shared_out, task_out)
                else:
                    diff_loss = torch.tensor(0).to(device=self.device, dtype=torch.float32)
                    self.diff_loss_reg = 0

                # Loss values
                task_loss = self.task_loss(output, y)

                total_loss = task_loss + self.diff_loss_reg * diff_loss

                loss_t += task_loss
                # loss_a += adv_loss
                loss_d += diff_loss
                loss_total += total_loss
                num += x.size(0)

        res['loss_t'], res['acc_t'] = loss_t.item() / (batch + 1), 100 * correct_t / num
        res['loss_d'] = loss_d.item() / (batch + 1)
        res['loss_tot'] = loss_total.item() / (batch + 1)
        res['size'] = self.loader_size(data_loader)

        return res
```

I see that you got also a CUDA device error. You can remove the device number given in the config file by changing `device: "cuda:7"` to `device: "cuda"`. 

Hope this helps. Let me know if you have any further questions otherwise please close the issue if it is resolved.

**Comment by drforester** [comment]
wow!, thank you for the quick and comprehensive answer. I'm way overdue for sleep or I'd try your solution now. I'll certainly let you know and close the issue soon.

**Comment by drforester** [comment]
The two functions work just like you said; thanks again! For each of my 20 tasks, I noticed that the collection of class Ids are different than they were at training...but I don't see how the that would matter to the trained model. However I am getting significantly lower accuracies during testing than I got after training the last model. For example, after training, using final_model_id = 19, I got:
![Screenshot at 2020-05-30 09-47-59](https://user-images.githubusercontent.com/60823216/83331421-f38c4000-a25b-11ea-860e-a76c6a1dc225.png)

and after testing the model just now  I get values like this:
![Screenshot at 2020-05-30 09-57-55](https://user-images.githubusercontent.com/60823216/83331442-0f8fe180-a25c-11ea-8474-dc1c8dc30bc1.png)

Any idea what I'm doing wrong?

**Comment by SaynaEbrahimi** [comment]
You must use the same `args.seed` as you used during training so that you get the same data splits for training and test data. Please see [these lines](https://github.com/facebookresearch/Adversarial-Continual-Learning/blob/4474bc02122998bb74202ac0bcc90cc0e5024f62/src/dataloaders/cifar100.py#L201) in the data generators for each dataset.

**Comment by drforester** [comment]
Thank you for your help! Using the same seed gives the expected results. I can't wait to try this on my own data!

**Comment by drforester** [comment]
Happy to say that I trained on my own image dataset and was able to use your inference method to get good results! The set consists of 6 classes and I trained them in 3 tasks of 2 classes each. When I use the inference method, it expects that I include the task_id. What if I wanted to pass in single images from a randomized collection, where the class-id (and task-id) are unknown to the user. Is it possible to do inference this way using ACL? Can it return a probabilities for class-ids instead of probabilities of task-ids?

**Comment by mipet23** [comment]
Sorry, but I still do not understand why you load the old private models for testing. Why is it not possible to use the current network for testing? Why are gradients also calculated on P_i for i=0,...,t-1 when training Task t? I cannot find an explanation for this in either the paper or the code. It would be very kind of you if you could give me a hint on this.
